File: C++/myLeetCode/Greedy/968. Binary Tree Cameras.cpp
// 968. Binary Tree Cameras
// https://leetcode.com/problems/binary-tree-cameras/

#include <iostream>
#include "../Binary_Tree/BinaryTreeUtils.hpp"

class Solution {
public:
    int minCameraCover(TreeNode<int>* root) {
        int result = 0;
        if (dfs(root, result) == 0) {
            result++;
        }
        return result;
    }

private:
    int dfs(TreeNode<int>* node, int& result) {
        if (node == nullptr) {
            return 2;
        }
        
        int left = dfs(node->left, result);
        int right = dfs(node->right, result);
        
        if (left == 0 || right == 0) {
            result++;
            return 1;
        }
        
        if (left == 1 || right == 1) {
            return 2;
        }
        
        return 0;
    }
};

int main() {
    TreeNode<int>* root = buildTree({0, 0, -1, 0, 0}, -1);
    std::cout << "Minimum number of cameras required: " << Solution().minCameraCover(root) << std::endl;
    
    TreeNode<int>* root2 = buildTree({0, 0, -1, 0, -1, 0, -1, -1, 0}, -1);
    std::cout << "Minimum number of cameras required: " << Solution().minCameraCover(root2) << std::endl;

    return 0;
}

File: C++/代码随想录/Greedy/968. Binary Tree Cameras.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def minCameraCover(root):
    def dfs(node):
        if not node:
            return 0, 0, float('inf')
        L = dfs(node.left)
        R = dfs(node.right)
        dp0 = L[1] + R[1]
        dp1 = min(L[2] + min(R[1], R[2]), R[2] + min(L[1], L[2]))
        dp2 = 1 + min(L) + min(R)
        return dp0, dp1, dp2

    return min(dfs(root)[1:])

# 测试用例
root = TreeNode(0, TreeNode(0, TreeNode(0), TreeNode(0)), None)
print(minCameraCover(root))  # 应该输出 1
